les/fetcher.go

```diff
@@ -68,46 +68,47 @@ type fetcherPeer struct {
 	latest *announceData // The latest announcement sent from the peer
 
 	// These following two fields can track the latest announces
-	// from the peer with limited size for caching.
+	// from the peer with limited size for caching. We hold the
+	// assumption that all enqueued announces are td-monotonic.
 	announces     map[common.Hash]*announce // Announcement map
-	announceQueue *prque.Prque              // Announcement queue
+	announcesList []common.Hash             // FIFO announces list
 }
 
 // addAnno enqueues an new trusted announcement. If the queued announces overflow,
 // evict from the oldest.
-func (fp *fetcherPeer) addAnno(announce *announce) {
-	// Short circuit if the announce already exists. In normal case it should
-	// never happen since only monotonic announce is accepted. But the adversary
+func (fp *fetcherPeer) addAnno(anno *announce) {
+	// Short circuit if the anno already exists. In normal case it should
+	// never happen since only monotonic anno is accepted. But the adversary
 	// may feed us fake announces with higher td but same hash. In this case,
-	// ignore the announce anyway.
-	hash, number := announce.data.Hash, announce.data.Number
+	// ignore the anno anyway.
+	hash := anno.data.Hash
 	if _, exist := fp.announces[hash]; exist {
 		return
 	}
-	fp.announces[hash] = announce
-	fp.announceQueue.Push(hash, -int64(number))
+	fp.announces[hash] = anno
+	fp.announcesList = append(fp.announcesList, hash)
 
 	// Evict oldest if the announces are oversized.
-	for fp.announceQueue.Size() > cachedAnnosThreshold {
-		item, _ := fp.announceQueue.Pop()
-		delete(fp.announces, item.(common.Hash))
+	for len(fp.announcesList) > cachedAnnosThreshold {
+		delete(fp.announces, fp.announcesList[0])
+		fp.announcesList = fp.announcesList[1:]
 	}
 }
 
 // forwardAnno removes all announces from the map with a number lower than
 // the provided threshold.
-func (fp *fetcherPeer) forwardAnno(number uint64) []*announce {
-	var ret []*announce
-	for !fp.announceQueue.Empty() {
-		item, priority := fp.announceQueue.Pop()
-		if uint64(-priority) > number {
-			fp.announceQueue.Push(item, priority)
+func (fp *fetcherPeer) forwardAnno(td *big.Int) []*announce {
+	var evicted []*announce
+	for len(fp.announcesList) > 0 {
+		anno := fp.announces[fp.announcesList[0]]
+		if anno.data.Td.Cmp(td) > 0 {
 			break
 		}
-		ret = append(ret, fp.announces[item.(common.Hash)])
-		delete(fp.announces, item.(common.Hash))
+		evicted = append(evicted, anno)
```

Comment:
Just a nitpick but instead of appending items to `evicted` one by one you could also just iterate the index until you find the first element that is going to stay, then cut the beginning of the slice in one step.

Reply:
fixed